File: daemon/yeelight/Controller.php
<?php


namespace Ething\Yeelight;


use \Ething\Yeelight\Yeelight;
use \Ething\Device\YeelightDevice;


class Controller extends \Stream {
	
	const AUTOCONNECT_PERIOD = 60; // seconds	
	const RESPONSE_TIMEOUT = 10; // seconds	
	const CONNECT_TIMEOUT = 2; // seconds	
	
	
	public $device = null;
	
	protected $logger;
	
	private $lastActivity = 0;
	
	private $buffer = "";
	
	private $lastAutoconnectLoop = 0;
	private $preventFailConnectLog = false;
	
	protected $isOpened = false;
	
	// response management
	private $responseListeners = array();
	
	public $options = array();
	
	public function __construct(YeelightDevice $device, array $options = array()){
		
		$this->device = $device;
		$this->options = array_replace_recursive($this->options, $options);
		$this->logger = $this->device->ething->logger();
	}
	
	public function ething(){
		return $this->device->ething;
	}
	
	public function name(){
		return $this->device->name();
	}
	
	public function __get($name){
		return isset($this->options[$name]) ? $this->options[$name] : null;
    }
	
	public function open(){
		
		if($this->isOpened){
			if(!$this->close())
				return false; // unable to close !
		}
		
		$device = $this->device;
		$host = $device->get('host');
		
		$stream = @stream_socket_client("tcp://".$host.':55443', $errno, $errstr, self::CONNECT_TIMEOUT);
		if($stream === false)
			throw new \Exception("Yeelight: unable to connect to the device {$host} : {$errstr}");
		
		// make this stream non blocking !
		stream_set_blocking($stream, false);
		
		$this->stream = $stream;
		$this->lastActivity = time();
		$this->buffer = '';
		
		$this->isOpened = true;
		$this->logger->info("Yeelight: connected to {$host}");
		
		$device->setConnectState(true);
		
		return true;
	}
	
	public function read(){
		if($this->isOpened){
			
			$chunk = fgets($this->stream);
			if($chunk===false){
				// an error occurs
				$this->close();
				return;
			}
			$this->buffer .= $chunk;
			
			$this->lastActivity = time();
			
			if(strlen($chunk) === 0){
				// connection closed
				$this->close();
				
			} else if(false !== ($p = strrpos($this->buffer, "\n")) ){
				
				$lines = preg_split("/\r?\n/", substr($this->buffer, 0, $p+1));
				$this->buffer = substr($this->buffer, $p+1);
				
				foreach($lines as $line){
					
					$line = trim($line);
					if(empty($line)) continue;
					
					try {
						// must be json
						$message = json_decode($line, true);
						if(!is_array($message)) throw new \Exception();
						
						$this->logger->debug("Yeelight: message received = {$line}");
						$this->processMessage($message);
						
					} catch (\Exception $e) {
						// skip the line
						$this->logger->warn("Yeelight: unable to handle the message {$line}");
						continue;
					}
					
				}
				
			}
		}
	}
	
	
	public function write($str){
		if($this->isOpened){
			$this->lastActivity = time();
			return @fwrite($this->stream, $str);
		}
		return 0;
	}
	
	public function close(){
		if( $this->isOpened ){
			@fclose($this->stream);
			$this->stream = null;
			$this->isOpened = false;
			$this->lastAutoconnectLoop = 0;
			$this->device->setConnectState(false);
			$this->logger->info("Yeelight: closed");
		}
		return !$this->isOpened;
	}
	
	/*
	exemple of messages : (https://www.yeelight.com/download/Yeelight_Inter-Operation_Spec.pdf)
		// response to the query {"id":1,"method":"get_prop","params":["power", "not_exist", "bright"]}
		{"id":1, "result":["off","","100"]}
		// notifications
		{"method":"props","params":{"power":"on"}}
		{"method":"props","params":{"power":"off"}}
		
	*/
	public function processMessage($message) {
		
		$device = $this->device;
		
		// refresh the device meta data before update it !
		$device->refresh();
		
		$device->updateSeenDate();
		
		$isResponse = isset($message["id"]);
		
		if($isResponse){
			
			$responseId = intval($message["id"]);
			$responseResult = isset($message["result"]) ? $message["result"] : array();
			
			foreach($this->responseListeners as $i => $responseListener){
				
				if($responseListener['id'] === $responseId){
					// remove this item
					array_splice($this->responseListeners, $i, 1);
					
					if(is_callable($responseListener['callback']))
						call_user_func($responseListener['callback'], false, $responseResult);
					
					break;
				}
			}
			
		} else if(isset($message["method"]) && isset($message["params"])) {
			// notification
			
			$method = $message["method"];
			$params = $message["params"];
			
			
			if($method === "props"){
				
				foreach($params as &$value){
					if(is_numeric($value))
						$value = $value + 0;
				}
				
				$device->storeData($params);
				
			} else {
				throw new \Exception();
			}
			
		} else {
			throw new \Exception();
		}
		
	}
	
	
	
	private $lastState_ = false;
	
	public function update(){
		// do some stuff regularly
		$now = microtime(true);
		
		// check for a deconnection
		if(!$this->isOpened && $this->isOpened != $this->lastState_)
			$this->logger->info("Yeelight: disconnected");
		$this->lastState_ = $this->isOpened;
		
		// autoconnect
		if(!$this->isOpened && ($now - $this->lastAutoconnectLoop) > self::AUTOCONNECT_PERIOD ){
			try{
				$this->open();
				$this->logger->info("Yeelight: connected");
				$this->preventFailConnectLog = false;
			} catch(\Exception $e){
				if(!$this->preventFailConnectLog) $this->logger->warn("Yeelight: unable to connect : {$e->getMessage()}");
				$this->preventFailConnectLog = true;
			}
			$this->lastAutoconnectLoop = $now;
		}
		
		// check response timeout
		foreach($this->responseListeners as $i => $responseListener){
			if( ($now - $responseListener['ts']) > self::RESPONSE_TIMEOUT ){
				
				// remove this item
				array_splice($this->responseListeners, $i, 1);
				
				if(is_callable($responseListener['callback'])){
					call_user_func($responseListener['callback'], 'response timeout', null);
				}
			}
		}
	}
	
	
	/*
	*  $message message to send
	*  $callback (optional) function($error, $messageSent, $messageReceived = null)
	*  $waitResponse (optional) true|false wait for a response or not
	*/
	public function send(array $message, $callback = null, $waitResponse = null) {
		
		$message['id'] = rand(1,9999);
		
		$this->logger->debug("Yeelight: message send id={$message['id']}");
		
		if(!$this->isOpened){
			if(is_callable($callback)){
				call_user_func($callback, 'not connected', $message);
			}
			return 0;
		}
		
		$wb =  $this->write(json_encode($message)."\r\n");
		
		if($waitResponse){
			// wait for a response
			$this->responseListeners[] = array(
				'callback' => function($error, $messageReceived) use($callback, $message) {
					if(is_callable($callback)) call_user_func($callback, $error, $message, $messageReceived);
				},
				'ts' => microtime(true),
				'messageSent' => $message,
				'id' => $message['id']
			);
		} else {
			if(is_callable($callback)) call_user_func($callback, false, $message);
		}
		
		return $wb;
	}
	
	protected $stream = null;
	
	public function getStreams(){
		return array($this->stream);
	}
	
	public function process($stream){
		$this->read();
	}
	
};

File: ething/yeelight/Controller.py


import yeelight as Yeelight
import socket
import time
import json
import random


class Controller (object):
	
	AUTOCONNECT_PERIOD = 60; # seconds	
	RESPONSE_TIMEOUT = 10; # seconds	
	CONNECT_TIMEOUT = 2; # seconds	
	
	reset_attr = ['host']
	
	
	
	def __init__ (self, device):
		
		self._device = device
		self._ething = device.ething
		self._isOpened = False
		
		self._log = device.ething.log
		
		self._sock = None
		
		self._lastState = False
		self._lastAutoconnectLoop = 0
		self._preventFailConnectLog = 0
		self._lastActivity = 0
		
		self._buffer = ""
		
		# response management
		self._responseListeners = []
		
		# refresh the gateway each time the gateway properties changes
		self.ething.signalManager.bind('ResourceMetaUpdated', self.onResourceMetaUpdated)
		
		self.ething.scheduler.tick(self.update)
	
	
	@property
	def device (self):
		return self._device
	
	@property
	def log (self):
		return self._log
	
	@property
	def ething (self):
		return self._ething
	
	@property
	def isOpened (self):
		return self._sock is not None
	
	@property
	def socket (self):
		return self._sock
	
	def onResourceMetaUpdated(self, signal):
		if signal['resource'] == self.device.id and signal['rModifiedDate'] > self.device.modifiedDate :
			self.device.refresh()
			
			for attr in signal['attributes']:
				if attr in Controller.reset_attr:
					self.restart()
					break
	
	def destroy(self):
		self.close()
		self.ething.signalManager.unbind('ResourceMetaUpdated', self.onResourceMetaUpdated);
	
	
	def open (self):
		self._lastAutoconnectLoop = time.time()
		
		if self.isOpened:
			return True
		
		host = self.device.host
		
		self._sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
		self._sock.settimeout(Controller.CONNECT_TIMEOUT)
		
		try:
			self._sock.connect((host, Yeelight.PORT))
		except socket.error as e:
			raise Exception("Yeelight: unable to connect to the device %s : %s" % (host,str(e)))
		
		self._sock.setblocking(0)
		
		self.ething.socketManager.registerReadSocket(self._sock, self.onDataAvailable)
		
		self._lastActivity = time.time()
		self._buffer = ''
		
		self.log.info("Yeelight: connected to %s" % host)
		
		self.device.setConnectState(True)
		
		return True
	
	def readline(self):
		p = self._buffer.find("\n")
		if p>=0 :
			line = self._buffer[0:p]
			self._buffer = self._buffer[p+1:]
			return line.rstrip()
		
		return None
	
	def onDataAvailable (self):
		
		
		chunk = self._sock.recv(1024)
		if not chunk:
			# connection broken
			self.close()
			return
		
		self._buffer += chunk
		
		self._lastActivity = time.time()
		
		
		while True:
			line = self.readline()
			if line is not None:
				
				self.log.debug("Yeelight: message received = %s" % line)
				
				try:
					
					# must be json
					message = json.loads(line)
					
					self.processMessage(message)
					
				except Exception as e:
					# skip the line
					self.log.exception("Yeelight: unable to handle the message %s" % (line))
					continue
			else:
				break
	
	
	
	
	
	
	def close (self):
		if self.isOpened:
			self.ething.socketManager.unregisterReadSocket(self._sock)
			self._sock.close()
			self._sock = None
			
			self.device.setConnectState(False)
			
			for responseListener in self._responseListeners:
				responseListener['callback']('disconnected', None);
			self._responseListeners = []
			
			self.log.info("Yeelight: closed")
			
	
	
	"""
	exemple of messages : (https://www.yeelight.com/download/Yeelight_Inter-Operation_Spec.pdf)
		// response to the query {"id":1,"method":"get_prop","params":["power", "not_exist", "bright"]}
		{"id":1, "result":["off","","100"]}
		// notifications
		{"method":"props","params":{"power":"on"}}
		{"method":"props","params":{"power":"off"}}
		
	"""
	def processMessage (self, message):
		
		with self._device as device:
			
			device.setConnectState(True)
			
			if "id" in message:
				
				responseId = int(message["id"])
				responseResult = message["result"] if "result" in message else []
				
				i = 0
				while i < len(self._responseListeners):
					responseListener = self._responseListeners[i]
					
					if responseListener['id'] == responseId:
						
						# remove this item
						self._responseListeners.pop(i)
						i -= 1
						
						if callable(responseListener['callback']):
							responseListener['callback'](False, responseResult)
						
						break
					
					i += 1
				
			elif ("method" in message) and ("params" in message):
				# notification
				
				method = message["method"]
				params = message["params"]
				
				if method == "props":
					if params:
						device.storeData(params)
					
				else:
					raise Exception('unknown method %s' % str(method))
				
				
			else:
				raise Exception('unable to parse the message')
		
	
	
	
	# Retrieve and return the properties of the device.
	def refresh (self):
		
		device = self._device
		
		requestedProperties = [
            "power", "bright", "ct", "rgb", "hue", "sat",
            "color_mode", "flowing", "delayoff", "flow_params",
            "music_on", "name"
        ]
		
		def cb(error, messageSent, response):
			
			if not error and isinstance(response, list) and len(response) == len(requestedProperties):
				
				# some formatting
				for i in range(0, len(response)):
					if response[i].decode('utf8').isnumeric():
						response[i] = int(response[i])
				
				params = dict(zip(requestedProperties, response))
				
				device.storeData(params)
		
		
		self.send({
			"method" : "get_prop",
			"params" : requestedProperties
		}, cb, True)
		
	
	
	
	def update (self):
		
		# do some stuff regularly
		now = time.time()
		
		# check for a deconnection
		if not self.isOpened and self.isOpened != self._lastState:
			self.log.info("Yeelight: disconnected")
		self._lastState = self.isOpened
		
		# autoconnect
		if not self.isOpened and (now - self._lastAutoconnectLoop) > Controller.AUTOCONNECT_PERIOD:
			try:
				self._lastAutoconnectLoop = now
				self.open()
				self._preventFailConnectLog = 0
			except Exception as e:
				
				if self._preventFailConnectLog % 20 == 0:
					self.log.warn("Yeelight: unable to connect : %s" % e.message)
				self._preventFailConnectLog += 1
		
		# check for timeout !
		i = 0
		while i < len(self._responseListeners):
			responseListener = self._responseListeners[i]
			
			if now - responseListener['ts'] > Controller.RESPONSE_TIMEOUT:
				
				# remove this item
				self._responseListeners.pop(i)
				i -= 1
				
				responseListener['callback']('response timeout', None)
			
			i += 1
		
	
	
	
	"""
	 $message message to send
	 $callback (optional) function($error, $messageSent, $messageReceived = null)
	 $waitResponse (optional) true|false wait for a response or not
	"""
	def send (self, message, callback = None, waitResponse = None):
		
		if self.isOpened:
			
			message['id'] = random.randint(1,9999)
			
			self.log.debug("Yeelight: message send %s" % str(message))
			
			self._lastActivity = time.time()
			
			wb = self._sock.send(json.dumps(message)+"\r\n")
			
			if waitResponse:
				# wait for a response
				
				def cb(error, messageReceived):
					if callable(callback):
						callback(error, message, messageReceived)
				
				self._responseListeners.append({
					'callback' : cb,
					'ts' : time.time(),
					'messageSent' : message,
					'id' : message['id']
				})
				
			else:
				if callable(callback):
					callback(False, message, None)
			
			return wb
		
		else:
			
			if callable(callback):
				callback('not connected', message, None)
			
			return 0
	
	



